File: sopack/pack.py
"""``book.json`` (+ the committed calibration fixture) → ``.sopack`` — the
slow, Mac-only step.

This is the **only** module in the pipeline that imports an embedding
library, and it imports it **at module import time**, not inside a function.
That is deliberate (R7 / failures #2, #3): ``import sopack.pack`` is what the
CLI does the instant someone runs ``sopack pack``, so a missing or
mismatched ``fastembed`` fails in the first second, not twenty minutes into a
run. ``sopack.cli`` imports this module lazily, only for the ``pack``
subcommand, so every other command stays fastembed-free.

Against the seam ``sopack.book`` (owned by another agent — not defined by
this module, only imported): ``Book``, ``Block``, ``load``, ``to_payload``,
``uid``, ``BookError``. This module never inspects ``Book``/``Block``
internals beyond iterating ``book.blocks`` — every piece of per-book metadata
it needs (``lang``, ``book_code``, ``title``, ``author``, ``year``,
``corpus``, ``slug``, ``book_pair``) is read back out of the **payload**
``to_payload`` returns, because the ``sop``/``bible`` contract already
requires those fields on every point. That keeps this module decoupled from
whatever shape ``Book`` turns out to have.
"""

from __future__ import annotations

import hashlib
import json
import platform
import secrets
from datetime import datetime, timezone
from pathlib import Path

import fastembed
from fastembed import TextEmbedding

from . import __version__ as SOPACK_VERSION
from . import contract
from .book import Book, Block, BookError, load, to_payload, uid  # noqa: F401  (seam)
from .format import PackWriter

__all__ = ["pack", "PackBuildError", "CalibrationFailed"]


class PackBuildError(Exception):
    """The pack could not be built — a contract violation or bad input."""


class CalibrationFailed(PackBuildError):
    """The pack-time calibration self-check scored below
    ``contract.PACK_MIN_COSINE`` (SOPACK-AUTONOMY.md §3.1). Raised before any
    book is embedded — a broken environment fails in seconds, not twenty
    minutes into a real run."""


# ── preflight (R7) ───────────────────────────────────────────────────────────

def _preflight() -> None:
    """Everything this run will eventually need, checked before any slow work.

    ``import fastembed`` already happened at module level (see the module
    docstring); what is left is asserting it is the *right* fastembed —
    version and, transitively, pooling. This pin is a property of THIS
    Python reference implementation, not of the neutral contract (a Rust
    build has no fastembed at all) — acceptance of the vectors themselves is
    decided by the calibration self-check below, not by the library version;
    this check only turns a silently-wrong pooling into a fast, readable
    failure instead of a self-check that fails anyway but says less.
    """
    got_version = getattr(fastembed, "__version__", None)
    want_version = contract.PYTHON_FASTEMBED_VERSION
    if got_version != want_version:
        raise PackBuildError(
            f"fastembed {got_version!r} is installed, this sopack build requires "
            f"{want_version!r} — pin the right version before packing. "
            "A different fastembed version can silently use different "
            "pooling, which writes vectors into a different geometric space "
            "than the live collection (this is failure #15).")


def _to_list(vector) -> list[float]:
    tolist = getattr(vector, "tolist", None)
    return tolist() if tolist is not None else list(vector)


def _attr(obj, name, default=None):
    if hasattr(obj, name):
        return getattr(obj, name)
    if isinstance(obj, dict):
        return obj.get(name, default)
    return default


def _load_book(item) -> tuple[object, str | None]:
    """*item* is either a path to a book.json, or a pre-loaded
    ``(book, sha256_or_None)`` pair (how tests hand in a fake Book without a
    real file, and how a caller can reuse an already-parsed Book)."""
    if isinstance(item, tuple) and len(item) == 2:
        return item
    path = Path(item)
    raw = path.read_bytes()
    book_obj = load(path)
    return book_obj, hashlib.sha256(raw).hexdigest()


def _default_pack_id(profile_name: str) -> str:
    stamp = datetime.now(timezone.utc).strftime("%Y-%m-%d")
    return f"{profile_name}-{stamp}-{secrets.token_hex(2)}"


def _default_created_by() -> str:
    return (f"sopack {SOPACK_VERSION} on "
            f"{platform.system()} {platform.release()} {platform.machine()}")


def _load_calibration(calibration) -> tuple[dict, str]:
    """Resolve *calibration* (the ``pack()`` parameter) into ``(fixture_doc,
    fixture_sha256)``.

    ``None`` (the normal case) loads and sha256-verifies the contract's own
    committed fixture via :func:`contract.load_calibration`; the sha is the
    contract's own ``CALIBRATION_SHA256``. A path, dict or list (tests: an
    explicit override) has no committed file to check against, so its sha is
    computed from its own canonical JSON (:func:`contract.calibration_fixture_sha256`)
    — the importer, given that SAME override, computes the identical sha, so
    the two sides still agree on what "this fixture" is."""
    if calibration is None:
        return contract.load_calibration(), contract.CALIBRATION_SHA256
    if isinstance(calibration, (str, Path)):
        try:
            raw = Path(calibration).read_bytes()
        except OSError as exc:
            raise PackBuildError(f"cannot read calibration fixture {calibration}: {exc}") from exc
        try:
            doc = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise PackBuildError(f"{calibration}: not valid JSON: {exc}") from exc
    elif isinstance(calibration, dict):
        doc = calibration
    else:
        doc = {"schema": contract.SCHEMA_CALIBRATION, "entries": list(calibration)}
    return doc, contract.calibration_fixture_sha256(doc)


def _titles_fragment(profile: contract.Profile, book_entries: list[dict]) -> dict | None:
    """Additive fragment matching ``app/data/sop_books.json``'s shape —
    ``{lang: {code: {"titles": [...], author?, year?, corpus?, en_code?}}}`` —
    so it can be merged with exactly the logic in
    ``scripts/merge_corpus_titles.py`` / ``export_book_titles.make_adder``.
    ``bible`` has no title table (see contract.py's profile comparison)."""
    if profile.name != "sop":
        return None
    frag: dict[str, dict[str, dict]] = {}
    for e in book_entries:
        lang, code = e["lang"], e["book_code"]
        entry = frag.setdefault(lang, {}).setdefault(code, {"titles": []})
        title = e.get("title")
        if title and title not in entry["titles"]:
            entry["titles"].append(title)
        if e.get("author") and not entry.get("author"):
            entry["author"] = e["author"]
        if e.get("year") and not entry.get("year"):
            entry["year"] = e["year"]
        if e.get("corpus") and not entry.get("corpus"):
            entry["corpus"] = e["corpus"]
        book_pair = e.get("book_pair") or ""
        if lang != "en" and "/" in book_pair:
            en_code = book_pair.split("/", 1)[1].strip()
            if en_code and not entry.get("en_code"):
                entry["en_code"] = en_code
    return frag


def pack(books, out_path, *, profile: str | None = None,
          pack_id: str | None = None, created_by: str | None = None,
          id_rule: str | None = None, batch_size: int = 1,
          workers: int | None = None, progress=print,
          calibration=None, device: str = "cpu", threads: int = 1) -> dict:
    """Embed every block of *books* and write ``out_path`` as a ``.sopack``.

    ``books`` — paths to ``book.json`` files (or, for tests / reuse,
    ``(book, sha256_or_None)`` pairs already loaded).
    ``calibration`` — ``None`` (the normal case: load and sha256-verify the
    contract's own committed fixture, ``contracts/<id>/calibration.json``), a
    path to an alternate fixture file, or an already-loaded fixture
    doc/entry-list (tests). See :func:`_load_calibration`.
    ``batch_size`` — defaults to **1**, not a larger number: M0 measured
    single-text batches at 2.73 blocks/s on CPU vs 0.99 blocks/s at batch 32
    (large batches are a GPU lever, not a CPU one — SOPACK-1.0-PLAN.md §2/§3.3).
    ``workers`` — ``fastembed`` parallel worker count; ``None``/``0`` (the
    default) means single-process — see the note above the workers-resolution
    line for why this does NOT default to ``os.cpu_count()``. When a caller
    does pass a count, it must come from ``os.cpu_count()``, never a shelled
    out ``nproc`` (R12/#4) — that part of the rule is unaffected.

    Returns the manifest actually written, read back from the finished file —
    so what this function returns is provably what is on disk, not merely
    what it intended to write.
    """
    _preflight()

    if not books:
        raise PackBuildError("no books given")

    loaded = [_load_book(item) for item in books]

    profile_names = {_attr(b, "profile") for b, _ in loaded}
    profile_names.discard(None)
    if profile is not None:
        if profile_names - {profile}:
            raise PackBuildError(
                f"--profile {profile!r} was given but book(s) declare "
                f"{sorted(profile_names - {profile})!r}")
        profile_name = profile
    elif len(profile_names) == 1:
        profile_name = next(iter(profile_names))
    elif not profile_names:
        raise PackBuildError(
            "no book declares a profile and none was given with --profile")
    else:
        raise PackBuildError(
            f"books declare different profiles {sorted(profile_names)!r} — "
            "pack them separately, one profile per .sopack")
    prof = contract.get_profile(profile_name)

    fixture, fixture_sha = _load_calibration(calibration)
    fixture_entries = fixture.get("entries") or []
    if not fixture_entries:
        raise PackBuildError(
            "calibration fixture has no entries — the self-check is not "
            "skippable (R11); see docs/SOPACK-AUTONOMY.md §3.1")

    if id_rule is not None:
        resolved_id_rule = id_rule
    else:
        declared = {_attr(b, "id_rule") for b, _ in loaded}
        declared.discard(None)
        if len(declared) > 1:
            raise PackBuildError(
                f"books declare different id_rules {sorted(declared)!r} in one "
                "pack — a .sopack carries exactly one id_rule; pack them "
                "separately or pass --id-rule to force one")
        resolved_id_rule = next(iter(declared), prof.default_id_rule)
    if resolved_id_rule not in prof.id_rules:
        raise PackBuildError(
            f"id_rule {resolved_id_rule!r} is not valid for profile "
            f"{prof.name!r} (allowed: {', '.join(prof.id_rules)})")

    pack_id = pack_id or _default_pack_id(prof.name)
    created_by = created_by or _default_created_by()
    # `workers` is deliberately NOT defaulted to os.cpu_count() here, even
    # though R12/#4 says a worker count must come from os.cpu_count() and
    # never `nproc` — that rule is about *how* a count is obtained when one is
    # wanted, not a mandate to always want one. fastembed's `parallel=N`
    # forks N worker subprocesses that EACH load their own copy of the ~2.2 GB
    # model; the parent process here has already loaded it once (this
    # function needs a single model instance for both books and canaries —
    # see the module docstring / R11), and forking after onnxruntime's
    # internal thread pool is already up is a known deadlock class. Measured
    # on this machine: `workers=os.cpu_count()` (4) against a real ~2500-block
    # book hung for 10 minutes and then failed with `_queue.Empty` — this
    # container's `MemAvailable` was ~1.9 GB, nowhere near 4x2.2 GB. Only use
    # multiple processes when a caller explicitly opts in with `--workers`,
    # on a machine known to have the RAM for it (the reference one-off's own
    # comment: "try (cores - 2) on a Mac").
    workers = workers or 0

    progress(f"loading {contract.EMBEDDING['model']} …")
    model = TextEmbedding(model_name=contract.EMBEDDING["model"])
    prefix = contract.EMBEDDING["passage_prefix"]
    embed_kwargs = {"batch_size": batch_size}
    if workers:
        embed_kwargs["parallel"] = workers

    # Calibration self-check (SOPACK-AUTONOMY.md §3.1, R11) — BEFORE any book
    # is embedded, with the same model instance and settings the books will
    # use. A broken environment (wrong pooling, a bad ORT build, …) then
    # fails in seconds instead of after however long the books take.
    progress(f"calibration self-check: embedding {len(fixture_entries)} "
             f"fixture entr{'y' if len(fixture_entries) == 1 else 'ies'} …")
    fixture_texts = [prefix + e["text"] for e in fixture_entries]
    fixture_vectors = [_to_list(v) for v in model.embed(fixture_texts, **embed_kwargs)]
    for v in fixture_vectors:
        if len(v) != contract.VECTOR_SIZE:
            raise PackBuildError(
                f"calibration embedding produced dimension {len(v)}, contract "
                f"requires {contract.VECTOR_SIZE}")
    cosines = [contract.cosine(fv, e["vector"]) for fv, e in zip(fixture_vectors, fixture_entries)]
    min_cosine = min(cosines)
    mean_cosine = sum(cosines) / len(cosines)
    if min_cosine < contract.PACK_MIN_COSINE:
        raise CalibrationFailed(
            f"calibration self-check FAILED: min cosine {min_cosine:.8f} < "
            f"{contract.PACK_MIN_COSINE} (mean {mean_cosine:.8f}, n={len(cosines)}) — "
            "this machine's embeddings do not reproduce the committed fixture's; "
            "refusing to embed any book. This is exactly the failure the "
            "calibration gate exists to catch (a pooling/library/model change "
            "silently moving vectors into a different geometric space) — see "
            "docs/SOPACK-AUTONOMY.md §3.1.")
    progress(f"  calibration ok: min cosine {min_cosine:.8f}, mean {mean_cosine:.8f} "
             f"(threshold {contract.PACK_MIN_COSINE})")
    self_check = {"n": len(cosines), "min_cosine": min_cosine, "mean_cosine": mean_cosine,
                  "threshold": contract.PACK_MIN_COSINE}

    out_path = Path(out_path)
    book_entries: list[dict] = []
    seen_dim: int | None = None

    with PackWriter(out_path, profile=prof.name, pack_id=pack_id,
                     created_by=created_by, id_rule=resolved_id_rule,
                     dim=contract.VECTOR_SIZE, device=device, threads=threads,
                     batch_tokens=batch_size) as writer:
        for book_obj, book_sha256 in loaded:
            blocks = _attr(book_obj, "blocks") or []
            blocks = list(blocks)
            if not blocks:
                progress("  (skipping book with no blocks)")
                continue

            payloads = [to_payload(book_obj, block) for block in blocks]
            uids = [uid(book_obj, block) for block in blocks]
            texts = [prefix + p[prof.text_field] for p in payloads]

            n_done = 0
            first_id = None
            book_label = payloads[0].get(prof.identity, "?")
            for start in range(0, len(texts), batch_size):
                chunk_texts = texts[start:start + batch_size]
                vectors = model.embed(chunk_texts, **embed_kwargs)
                for offset, vec in enumerate(vectors):
                    i = start + offset
                    vec = _to_list(vec)
                    if seen_dim is None:
                        seen_dim = len(vec)
                        if seen_dim != contract.VECTOR_SIZE:
                            raise PackBuildError(
                                f"embedding produced dimension {seen_dim}, "
                                f"contract requires {contract.VECTOR_SIZE}")
                    pid = writer.add(uids[i], payloads[i], vec)
                    if first_id is None:
                        first_id = pid
                    n_done += 1
                progress(f"  {book_label}: {n_done}/{len(texts)} blocks embedded "
                         f"| {writer.count} total")

            meta = payloads[0]
            book_entries.append({
                "book_code": meta.get(prof.identity, book_label),
                "lang": meta.get("lang"),
                "points": n_done,
                "first_id": first_id,
                "title": meta.get("title"),
                "author": meta.get("author"),
                "year": meta.get("year"),
                "corpus": meta.get("corpus"),
                "slug": meta.get("slug"),
                "book_pair": meta.get("book_pair"),
                "id_rule": resolved_id_rule,
                "book_sha256": book_sha256,
            })

        writer.set_books(book_entries)
        writer.set_probe(fixture_entries, fixture_vectors, self_check=self_check,
                         fixture_sha256=fixture_sha)
        titles = _titles_fragment(prof, book_entries)
        if titles is not None:
            writer.set_titles(titles)

    from .format import PackReader
    with PackReader(out_path) as reader:
        manifest = reader.manifest
    progress(f"wrote {out_path} — {manifest['counts']['points']:,} points, "
             f"{manifest['counts']['books']} books")
    return manifest

File: sopack-rs/crates/sopack-cli/src/commands/pack.rs
//! `sopack pack` — book.json(s) -> `.sopack` (`SOPACK-1.0-PLAN.md` §3.4/§3.6,
//! `SOPACK-2-FORMAT.md`). See the module docs below each phase for the
//! order this follows and how resume is implemented.

use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sopack_book::{to_payload, uid, Block, Book};
use sopack_format::{BookEntry, EmbeddingProvenance, PackWriter, ProbeEntryInput};
use sopack_progress::{ProgressSink, StageWeight, Unit};

use crate::args::{Cli, PackArgs};
use crate::contract_load::load_contract;
use crate::engine_build::build_engine;
use crate::exit::{CliError, EXIT_OK};
use crate::output::{print_json, print_text};
use crate::progress_build::build_progress;
use crate::provenance::{default_created_by, default_pack_id};

/// How many blocks are embedded and written between `PackWriter::checkpoint()`
/// calls — the resumability granularity. Small enough that
/// `SOPACK_TEST_ABORT_AFTER_BATCHES` gives a real crash-then-resume test on
/// even a small book, independent of the embedding engine's own internal
/// token-budget batching (`sopack_embed::batch`), which this is orthogonal
/// to: one CLI "batch" here can still turn into several ORT calls inside
/// `Engine::embed`.
const CHECKPOINT_BATCH_BLOCKS: usize = 16;

const FINGERPRINT_SCHEMA: &str = "sopack-cli.pack-resume/1";
const FINGERPRINT_FILE: &str = "cli-state.json";

#[derive(Debug, Serialize, Deserialize, PartialEq)]
struct PackFingerprint {
    schema: String,
    books: Vec<(String, String)>,
    contract_id: String,
    contract_sha256: String,
    profile: String,
    id_rule: String,
    device: String,
    threads: Option<usize>,
    batch_tokens: Option<usize>,
}

fn checkpoint_dir_for(out_path: &Path) -> PathBuf {
    let mut s = out_path.as_os_str().to_owned();
    s.push(".sopack.partial");
    PathBuf::from(s)
}

fn read_fingerprint(dir: &Path) -> Option<PackFingerprint> {
    let bytes = std::fs::read(dir.join(FINGERPRINT_FILE)).ok()?;
    serde_json::from_slice(&bytes).ok()
}

fn write_fingerprint(dir: &Path, fp: &PackFingerprint) -> Result<(), CliError> {
    let bytes = serde_json::to_vec_pretty(fp)?;
    std::fs::write(dir.join(FINGERPRINT_FILE), bytes)?;
    Ok(())
}

struct LoadedBook {
    path: PathBuf,
    sha256: String,
    book: Book,
}

fn load_books(paths: &[PathBuf]) -> Result<Vec<LoadedBook>, CliError> {
    if paths.is_empty() {
        return Err(CliError::usage("pack: no books given"));
    }
    let mut out = Vec::with_capacity(paths.len());
    for p in paths {
        let book = sopack_book::load(p)?;
        let problems = sopack_book::validate(&book);
        if !problems.is_empty() {
            return Err(CliError::input_invalid(format!(
                "{}: book.json is not valid, refusing to pack:\n  - {}",
                p.display(),
                problems.join("\n  - ")
            )));
        }
        let sha256 = sopack_book::book_sha256(p)?;
        out.push(LoadedBook {
            path: p.clone(),
            sha256,
            book,
        });
    }
    Ok(out)
}

fn resolve_profile(
    books: &[LoadedBook],
    given: Option<&str>,
) -> Result<&'static sopack_contract::Profile, CliError> {
    let declared: std::collections::BTreeSet<&str> =
        books.iter().map(|b| b.book.profile.as_str()).collect();
    let name = if let Some(p) = given {
        if declared.iter().any(|d| *d != p) {
            return Err(CliError::input_invalid(format!(
                "--profile {p:?} was given but book(s) declare {declared:?}"
            )));
        }
        p.to_string()
    } else if declared.len() == 1 {
        (*declared.iter().next().unwrap()).to_string()
    } else {
        return Err(CliError::input_invalid(format!(
            "books declare different profiles {declared:?} — pack them separately, one \
             profile per .sopack"
        )));
    };
    sopack_book::get_profile(&name).map_err(CliError::input_invalid)
}

fn resolve_id_rule(
    books: &[LoadedBook],
    profile: &sopack_contract::Profile,
    given: Option<&str>,
) -> Result<String, CliError> {
    let rule = if let Some(r) = given {
        r.to_string()
    } else {
        let declared: std::collections::BTreeSet<&str> =
            books.iter().map(|b| b.book.id_rule.as_str()).collect();
        if declared.len() > 1 {
            return Err(CliError::input_invalid(format!(
                "books declare different id_rules {declared:?} in one pack — a .sopack \
                 carries exactly one id_rule; pack them separately or pass --id-rule"
            )));
        }
        declared
            .into_iter()
            .next()
            .unwrap_or(&profile.default_id_rule)
            .to_string()
    };
    if !profile.id_rules.iter().any(|r| r == &rule) {
        return Err(CliError::input_invalid(format!(
            "id_rule {rule:?} is not valid for profile {:?} (allowed: {:?})",
            profile.name, profile.id_rules
        )));
    }
    Ok(rule)
}

/// Additive `titles.json` fragment — `sopack.pack._titles_fragment`. `bible`
/// has no title table.
fn titles_fragment(profile_name: &str, book_entries: &[BookEntry]) -> Option<serde_json::Value> {
    if profile_name != "sop" {
        return None;
    }
    let mut frag = serde_json::Map::new();
    for e in book_entries {
        let Some(lang) = &e.lang else { continue };
        let lang_map = frag
            .entry(lang.clone())
            .or_insert_with(|| serde_json::Value::Object(serde_json::Map::new()));
        let lang_obj = lang_map.as_object_mut().unwrap();
        let entry = lang_obj
            .entry(e.book_code.clone())
            .or_insert_with(|| serde_json::json!({"titles": []}));
        let entry_obj = entry.as_object_mut().unwrap();
        if let Some(title) = &e.title {
            let titles = entry_obj
                .get_mut("titles")
                .and_then(|v| v.as_array_mut())
                .unwrap();
            if !titles.iter().any(|t| t.as_str() == Some(title.as_str())) {
                titles.push(serde_json::Value::String(title.clone()));
            }
        }
        if let Some(author) = &e.author {
            entry_obj
                .entry("author")
                .or_insert_with(|| serde_json::Value::String(author.clone()));
        }
        if let Some(year) = e.year {
            entry_obj
                .entry("year")
                .or_insert_with(|| serde_json::Value::from(year));
        }
        if let Some(corpus) = &e.corpus {
            entry_obj
                .entry("corpus")
                .or_insert_with(|| serde_json::Value::String(corpus.clone()));
        }
        if lang != "en" {
            if let Some(pair) = &e.book_pair {
                if let Some((_, en_code)) = pair.split_once('/') {
                    let en_code = en_code.trim();
                    if !en_code.is_empty() {
                        entry_obj
                            .entry("en_code")
                            .or_insert_with(|| serde_json::Value::String(en_code.to_string()));
                    }
                }
            }
        }
    }
    Some(serde_json::Value::Object(frag))
}

#[derive(Serialize)]
struct PackResult {
    out: String,
    pack_id: String,
    profile: String,
    id_rule: String,
    points: u64,
    books: u64,
    device: String,
    calibration_min_cosine: f64,
    calibration_mean_cosine: f64,
    resumed: bool,
}

pub fn run(cli: &Cli, args: &PackArgs) -> Result<i32, CliError> {
    let contract = load_contract(&cli.contract)?;
    let loaded = load_books(&args.books)?;
    let profile = resolve_profile(&loaded, args.profile.as_deref())?;
    let id_rule = resolve_id_rule(&loaded, profile, args.id_rule.as_deref())?;
    let device_str = args.engine.device.clone();
    // Validated again inside build_engine; validated here too so a bad
    // --device value is reported before any checkpoint directory is touched.
    let _: sopack_embed::device::Device = device_str.parse().map_err(CliError::usage)?;

    let fingerprint = PackFingerprint {
        schema: FINGERPRINT_SCHEMA.to_string(),
        books: loaded
            .iter()
            .map(|b| (b.path.display().to_string(), b.sha256.clone()))
            .collect(),
        contract_id: contract.id.clone(),
        contract_sha256: contract.contract_sha256(),
        profile: profile.name.clone(),
        id_rule: id_rule.clone(),
        device: device_str.clone(),
        threads: args.engine.threads,
        batch_tokens: args.engine.batch_tokens,
    };

    let checkpoint_dir = checkpoint_dir_for(&args.out);
    let mut resumed = false;
    if checkpoint_dir.exists() {
        if args.fresh {
            std::fs::remove_dir_all(&checkpoint_dir)?;
        } else {
            match read_fingerprint(&checkpoint_dir) {
                Some(existing) if existing == fingerprint => resumed = true,
                Some(_) => {
                    return Err(CliError::input_invalid(format!(
                        "a checkpoint at {} was started with different inputs/options — \
                         re-run with --fresh to discard it and start over",
                        checkpoint_dir.display()
                    )))
                }
                None => {
                    return Err(CliError::input_invalid(format!(
                        "{} exists but has no readable {} — re-run with --fresh",
                        checkpoint_dir.display(),
                        FINGERPRINT_FILE
                    )))
                }
            }
        }
    }

    let progress = build_progress(
        cli.progress,
        vec![
            StageWeight::new("verify", 1.0),
            StageWeight::new("load_model", 1.0),
            StageWeight::new("calibrate", 1.0),
            StageWeight::new("embed", 10.0),
            StageWeight::new("write", 1.0),
        ],
    );

    let (mut engine, device_report, _model_dir) =
        build_engine(&contract, &args.engine, progress.as_ref())?;

    let fixture = contract.fixture.as_ref().ok_or_else(|| {
        CliError::resources(format!(
            "contract {:?} has no calibration fixture loaded",
            contract.id
        ))
    })?;
    let embed_fixture: sopack_embed::calibration::CalibrationFixture = fixture.into();
    let calib = match device_report.calibration {
        Some(c) => c,
        None => engine.calibrate(&embed_fixture, progress.as_ref())?,
    };
    let threshold = contract.calibration.pack_min_cosine;
    if !calib.passes(threshold) {
        return Err(CliError::calibration_failed(format!(
            "calibration self-check FAILED: min cosine {:.8} < {threshold} (mean {:.8}, n={}) \
             — refusing to embed any book",
            calib.min_cosine, calib.mean_cosine, calib.n
        )));
    }

    let provenance = EmbeddingProvenance {
        runtime: engine.runtime_string(env!("CARGO_PKG_VERSION")),
        device: engine.device().to_string(),
        threads: engine.threads() as u32,
        batch_tokens: engine.batch_tokens() as u32,
    };

    let mut writer = if resumed {
        PackWriter::resume(&args.out, &contract)?
    } else {
        let w = PackWriter::create(
            &args.out,
            &contract,
            &profile.name,
            args.pack_id
                .clone()
                .unwrap_or_else(|| default_pack_id(&profile.name)),
            args.created_by.clone().unwrap_or_else(default_created_by),
            Some(id_rule.clone()),
            provenance,
        )?;
        write_fingerprint(&checkpoint_dir, &fingerprint)?;
        w
    };

    // Flatten every (book, block) pair, in the order books/blocks were
    // given — resuming skips exactly `writer.count()` of them, matching how
    // many were durably appended before the crash.
    let mut flat: Vec<(usize, usize)> = Vec::new();
    for (bi, lb) in loaded.iter().enumerate() {
        if lb.book.blocks.is_empty() {
            progress.warn(&format!("{}: no blocks, skipping", lb.path.display()));
            continue;
        }
        for i in 0..lb.book.blocks.len() {
            flat.push((bi, i));
        }
    }
    let already = writer.count() as usize;
    if already > flat.len() {
        return Err(CliError::input_invalid(format!(
            "checkpoint at {} already holds {already} points, but only {} are expected from \
             these books — refusing to resume; use --fresh",
            checkpoint_dir.display(),
            flat.len()
        )));
    }

    // ONE "embed" stage for the whole run, sized in real subword tokens of
    // exactly the blocks still to embed (a resumed run only counts what is
    // left). Each checkpoint chunk below calls `engine.embed()`, which would
    // otherwise open and close its OWN "embed" stage per chunk — the overall
    // percentage then plateaued for the whole embed phase of any book over
    // one chunk (bench/README.md "Progress engine finding"). `AdvanceOnly`
    // forwards the chunk's token advances into this stage and swallows its
    // stage boundaries.
    let remaining_texts: Vec<String> = flat[already..]
        .iter()
        .map(|&(bi, ei)| block_text(&loaded[bi].book, &loaded[bi].book.blocks[ei], profile))
        .collect::<Result<_, CliError>>()?;
    let embed_total = engine.count_tokens(&remaining_texts)?;
    drop(remaining_texts);
    progress.stage_start("embed", embed_total, Unit::Tokens);
    let chunk_progress = AdvanceOnly(progress.as_ref());

    let abort_after: Option<usize> = std::env::var("SOPACK_TEST_ABORT_AFTER_BATCHES")
        .ok()
        .and_then(|v| v.parse().ok());
    let mut batches_done = 0usize;

    for chunk in flat[already..].chunks(CHECKPOINT_BATCH_BLOCKS) {
        let mut texts = Vec::with_capacity(chunk.len());
        let mut payloads = Vec::with_capacity(chunk.len());
        let mut uids = Vec::with_capacity(chunk.len());
        for &(bi, ei) in chunk {
            let book = &loaded[bi].book;
            let block: &Block = &book.blocks[ei];
            let payload = to_payload(book, block)?;
            let u = uid(book, block)?;
            let text = payload
                .get(profile.text_field.as_str())
                .and_then(|v| v.as_str())
                .unwrap_or_default()
                .to_string();
            texts.push(text);
            payloads.push(payload);
            uids.push(u);
        }
        let vectors = engine.embed(&texts, sopack_embed::spec::Role::Passage, &chunk_progress)?;
        for (i, vector) in vectors.iter().enumerate() {
            writer.add(&uids[i], payloads[i].clone(), vector)?;
        }
        writer.checkpoint()?;
        batches_done += 1;

        if let Some(n) = abort_after {
            if batches_done >= n {
                eprintln!(
                    "SOPACK_TEST_ABORT_AFTER_BATCHES={n}: simulating a crash after {batches_done} \
                     batch(es) — {} point(s) written, no finish()",
                    writer.count()
                );
                std::process::exit(crate::exit::CliErrorCode::Interrupted.exit());
            }
        }
    }
    progress.stage_end();

    let mut book_entries = Vec::new();
    for lb in loaded.iter() {
        if lb.book.blocks.is_empty() {
            continue;
        }
        let meta = &lb.book.book;
        let get = |k: &str| -> Option<String> {
            meta.get(k).and_then(|v| v.as_str()).map(str::to_string)
        };
        let get_year = || -> Option<i64> { meta.get("year").and_then(|v| v.as_i64()) };
        // Computed independently of how much of this run was resumed (see
        // the module doc): `sopack_book::uid`'s pre-hash string, uuid5'd the
        // same way `PackWriter::add` derives a point's id internally, so
        // the first point's id is always this book's first_id — not
        // whichever point happened to be embedded first *in this process*.
        let first_id = uid(&lb.book, &lb.book.blocks[0])
            .ok()
            .map(|u| uuid::Uuid::new_v5(&uuid::Uuid::NAMESPACE_DNS, u.as_bytes()).to_string());
        book_entries.push(BookEntry {
            book_code: get("book_code").unwrap_or_default(),
            lang: get("lang"),
            points: lb.book.blocks.len() as u64,
            first_id,
            title: get("title"),
            author: get("author"),
            year: get_year(),
            corpus: get("corpus"),
            slug: get("slug"),
            book_pair: get("book_pair"),
            id_rule: id_rule.clone(),
            book_sha256: Some(lb.sha256.clone()),
        });
    }

    let probe_entries: Vec<ProbeEntryInput> = fixture
        .entries
        .iter()
        .map(|e| ProbeEntryInput {
            id: e.id.clone(),
            profile: e.profile.clone(),
        })
        .collect();

    let pack_id = writer.pack_id().to_string();
    writer.set_books(book_entries.clone());
    writer.set_titles(titles_fragment(&profile.name, &book_entries));
    writer.set_probe(
        probe_entries,
        calib.probe_vectors.clone(),
        calib.n as u64,
        calib.min_cosine,
        calib.mean_cosine,
    )?;

    progress.stage_start("write", 1, Unit::Items);
    let out_path = writer.finish()?;
    progress.advance(1);
    progress.stage_end();
    progress.done();

    let result = PackResult {
        out: out_path.display().to_string(),
        pack_id,
        profile: profile.name.clone(),
        id_rule,
        points: book_entries.iter().map(|b| b.points).sum(),
        books: book_entries.len() as u64,
        device: engine.device().to_string(),
        calibration_min_cosine: calib.min_cosine,
        calibration_mean_cosine: calib.mean_cosine,
        resumed,
    };

    if cli.json {
        print_json(&serde_json::to_value(&result).unwrap());
    } else if !cli.quiet {
        print_text(&format!(
            "wrote {} — {} points, {} books (device={}, min_cosine={:.6})",
            result.out, result.points, result.books, result.device, result.calibration_min_cosine
        ));
    }
    Ok(EXIT_OK)
}

/// The text a block is embedded from: the profile's text field of its payload
/// (the same value the chunk loop embeds).
fn block_text(
    book: &sopack_book::Book,
    block: &Block,
    profile: &sopack_contract::Profile,
) -> Result<String, CliError> {
    let payload = to_payload(book, block)?;
    Ok(payload
        .get(profile.text_field.as_str())
        .and_then(|v| v.as_str())
        .unwrap_or_default()
        .to_string())
}

/// Forwards token advances and warnings into the caller's already-open
/// "embed" stage, and ignores the per-call stage boundaries `Engine::embed`
/// emits — so one stage spans every checkpoint chunk.
struct AdvanceOnly<'a>(&'a dyn ProgressSink);

impl ProgressSink for AdvanceOnly<'_> {
    fn stage_start(&self, _stage: &str, _total: u64, _unit: Unit) {}
    fn advance(&self, n: u64) {
        self.0.advance(n);
    }
    fn warn(&self, msg: &str) {
        self.0.warn(msg);
    }
    fn stage_end(&self) {}
    fn done(&self) {}
}
